client/json-rpc/src/async_client/client.rs

```diff
@@ -280,53 +376,61 @@ impl<R: RetryStrategy> Client<R> {
+        let rpc_resp: JsonRpcResponse = self.send_json_request(request.to_json(1)).await?;
+        self.validate(rpc_resp)
+    }
+
+    pub async fn batch_send(
         &self,
-        method: &str,
-        params: &serde_json::Value,
-    ) -> Result<JsonRpcResponse, Error> {
+        requests: Vec<Request>,
+    ) -> Result<Vec<Result<JsonRpcResponse, Error>>, Error> {
+        let json_array: Vec<serde_json::Value> = requests
+            .iter()
+            .enumerate()
+            .map(|(i, r)| r.to_json(i))
+            .collect();
+
+        let rpc_resp: Vec<JsonRpcResponse> = self.send_json_request(json![json_array]).await?;
+        return Ok(rpc_resp.iter().map(|r| self.validate(r.clone())).collect());
+    }
+
+    async fn send_json_request<T: for<'de> serde::Deserialize<'de>>(
+        &self,
+        req: serde_json::Value,
+    ) -> Result<T, Error> {
         let resp = self
             .http_client
             .post(self.server_url.clone())
-            .json(&json!({"jsonrpc": "2.0", "method": method, "params": params, "id": 1}))
+            .json(&req)
             .send()
             .await
             .map_err(Error::NetworkError)?;
         if resp.status() != 200 {
             return Err(Error::InvalidHTTPStatus(format!("{:#?}", resp)));
         }
-        let rpc_resp: JsonRpcResponse = resp.json().await.map_err(Error::InvalidHTTPResponse)?;
+        resp.json().await.map_err(Error::InvalidHTTPResponse)
+    }
 
-        if rpc_resp.jsonrpc != "2.0" {
-            return Err(Error::InvalidRpcResponse(rpc_resp));
+    fn validate(&self, resp: JsonRpcResponse) -> Result<JsonRpcResponse, Error> {
+        if resp.jsonrpc != "2.0" {
+            return Err(Error::InvalidRpcResponse(resp));
         }
 
         if let Some(state) = self.last_known_state() {
-            if rpc_resp.libra_chain_id != state.chain_id {
-                return Err(Error::ChainIdMismatch(rpc_resp));
+            if resp.libra_chain_id != state.chain_id {
+                return Err(Error::ChainIdMismatch(resp));
             }
         }
 
-        let resp_state = State::from_response(&rpc_resp);
+        let resp_state = State::from_response(&resp);
         if !self.update_state(resp_state) {
-            return Err(Error::StaleResponseError(rpc_resp));
+            return Err(Error::StaleResponseError(resp));
         }
 
-        if let Some(err) = rpc_resp.error {
+        if let Some(err) = resp.error {
             return Err(Error::JsonRpcError(err));
         }
-
-        Ok(rpc_resp)
-    }
-
-    pub fn update_state(&self, resp_state: State) -> bool {
-        let mut state_writer = self.last_known_state.write().unwrap();
-        if let Some(state) = &*state_writer {
-            if &resp_state < state {
-                return false;
-            }
-        }
-        *state_writer = Some(resp_state);
-        true
+        Ok(resp)
```

Comment:
I think you should add an req_id: id  here so you can verify  the id is same  for request and response. 

it is part of the jsonrpc protocol.

but in this case, if this client always send out id  0-X  for request in an batch you could also have an  "batch_validate()" wrapper that verify the response in an batch ,  I think you can also check to make sure responses all have same libra_timstamp  and libra_version etc , which may prevent bug from server side to affect client

Reply:
updated: batch will check id, as response order may not match requests order.
for non-batch call, still ignores response id for now.

Reply:
validate don't need response id, as you find response by request id, hence once you know a response for a request in batch case, no need validate response id.